### PR TITLE
[AutoDiff] Use ASTMangler for generated AD associated functions.

### PR DESCRIPTION
**Pull request by dan-zheng**
Use ASTMangler for generated AD associated functions:
JVPs, VJPs, differentials, pullbacks.

This centralizes the ad-hoc mangling logic scattered throughout the codebase.

Todos:
- [TF-680](https://bugs.swift.org/browse/TF-680): Mangle `@differentiable` attribute requirements.
- [TF-685](https://bugs.swift.org/browse/TF-685): Use ASTMangler for AD-related thunks.
- [TF-686](https://bugs.swift.org/browse/TF-686): Make Demangler/Remangler work with AD-generated functions.

### CODE REVIEW

**Comment by dan-zheng** [comment]
cc @bartchr808 
This PR will cause conflicts with https://github.com/apple/swift/pull/26057, I'm happy to help resolve them together.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/AST/ASTMangler.h [hunk 1]
```diff
@@ -151,7 +151,17 @@ class ASTMangler : public Mangler {
...
                                              ModuleDecl *Module);
-  
+
+  // SWIFT_ENABLE_TENSORFLOW
+  // Mangle the autodiff associated function with the given:
+  // - Mangled original function name.
+  // - Associated function kind.
+  // - Parameter/result indices.
+  // - Flag representing whether the function is a differential/pullback.
+  std::string mangleAutoDiffAssociatedFunctionHelper(
```
Comment (rxwei) [inline include/swift/AST/ASTMangler.h]: Associated function kind is orthogonal to linear maps. I’d suggest adding to separate mangling helpers, one for associates functions and one for linear maps. I think you should also add a LinearMapKind enum to distinguish between differential and pullback.

File: lib/AST/ASTMangler.cpp [hunk 2]
```diff
@@ -369,6 +369,37 @@ std::string ASTMangler::mangleReabstractionThunkHelper(
...
+std::string ASTMangler::mangleAutoDiffAssociatedFunctionHelper(
+    StringRef name, AutoDiffAssociatedFunctionKind kind,
+    const SILAutoDiffIndices &indices, bool isLinearMap) {
+  // TODO(TF-20): Make the mangling scheme robust.
+  // TODO(TF-680): Mangle `@differentiable` atttribute requirements as well.
+  beginManglingWithoutPrefix();
+
+  Buffer << "AD__";
+  Buffer << name;
+  Buffer << '_';
```
Comment (bartchr808) [inline lib/AST/ASTMangler.cpp]: Could these three lines be combined into: `Buffer << "AD__" << name << '_';`

File: include/swift/AST/ASTMangler.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/ASTMangler.h]: Added `AutoDiffLinearMapKind` in https://github.com/apple/swift/pull/26389/commits/221ddb0d76529c6f043b98af2e3734bdb5682be9. Let me know if you prefer the name `LinearMapKind`.

Added `ASTMangler::mangleAutoDiffLinearMapHelper` in https://github.com/apple/swift/pull/26389/commits/c5b2018a91fe40a34c44d4e7a164fc96f19d34b9.

File: lib/AST/ASTMangler.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/AST/ASTMangler.cpp]: Done in https://github.com/apple/swift/pull/26389/commits/c5b2018a91fe40a34c44d4e7a164fc96f19d34b9.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/AST/ASTMangler.h [hunk 3]
```diff
@@ -167,7 +167,7 @@ class ASTMangler : public Mangler {
   // - Associated function kind.
```
Comment (rxwei) [inline include/swift/AST/ASTMangler.h]: Linear map kind.

Comment (dan-zheng) [reply include/swift/AST/ASTMangler.h]: Done in https://github.com/apple/swift/pull/26389/commits/724aa0cf67c5902d95b8ef8b362ece8cba96b664.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**